velox/functions/prestosql/ArrayShuffle.cpp

```diff
@@ -0,0 +1,167 @@
+ * limitations under the License.
+ */
+#include <numeric>
+#include <random>
+#include "velox/expression/EvalCtx.h"
+#include "velox/expression/Expr.h"
+#include "velox/expression/VectorFunction.h"
+#include "velox/expression/VectorReaders.h"
+#include "velox/functions/lib/RowsTranslationUtil.h"
+
+namespace facebook::velox::functions {
+namespace {
+// See documentation at
+// https://prestodb.io/docs/current/functions/array.html#shuffle
+template <TypeKind T>
+class ArrayShuffleFunction : public exec::VectorFunction {
+ public:
+  ArrayShuffleFunction() {}
+
+  bool isDeterministic() const override {
+    return false;
+  }
+
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& /*outputType*/,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    VELOX_CHECK_EQ(args.size(), 1);
+    auto& arg = args[0];
+    VectorPtr localResult;
+
+    // Input can be constant, flat or dictionary.
+    if (arg->isConstantEncoding()) {
+      auto* constantArray = arg->as<ConstantVector<ComplexType>>();
+      const auto& flatArray = constantArray->valueVector();
+      const auto flatIndex = constantArray->index();
+
+      exec::LocalSelectivityVector singleRow(context, flatIndex + 1);
+      singleRow->setValid(flatIndex, true);
+      singleRow->updateBounds();
+
+      localResult = applyFlat(
+          *singleRow.get(), flatArray.get()->as<ArrayVector>(), context);
+      localResult =
+          BaseVector::wrapInConstant(rows.size(), flatIndex, localResult);
+    } else if (arg->isFlatEncoding()) {
+      localResult = applyFlat(rows, arg->as<ArrayVector>(), context);
+    } else { // dictionary encoding
+      exec::LocalDecodedVector decodedVectorHolder(context, *arg, rows);
+      localResult = applyFlat(
+          rows, decodedVectorHolder->base()->as<ArrayVector>(), context);
+    }
+
+    context.moveOrCopyResult(localResult, rows, result);
+  }
+
+  static std::vector<std::shared_ptr<exec::FunctionSignature>> signatures() {
+    return {// array(T) -> array(T)
+            exec::FunctionSignatureBuilder()
+                .typeVariable("T")
+                .returnType("array(T)")
+                .argumentType("array(T)")
+                .build()};
+  }
+
+ private:
+  VectorPtr applyFlat(
+      const SelectivityVector& rows,
+      const ArrayVector* inputArray,
+      exec::EvalCtx& context) const {
+    VectorPtr resultElements;
+
+    shuffleElementIndices(rows, inputArray, context, resultElements);
+
+    return std::make_shared<ArrayVector>(
+        context.pool(),
+        inputArray->type(),
+        inputArray->nulls(),
+        rows.end(),
+        inputArray->offsets(),
+        inputArray->sizes(),
+        resultElements,
+        inputArray->getNullCount());
+  }
+
+  void shuffleElementIndices(
+      const SelectivityVector& rows,
+      const ArrayVector* inputArray,
+      exec::EvalCtx& context,
+      VectorPtr& resultElements) const {
```

Comment:
The signature of this method would be clearer resultElements was returned via 'return'.

```
VectorPtr shuffleElementIndices(rows, inputArray, context)
```